go/mysqlconn/client.go

```diff
@@ -0,0 +1,455 @@
+package mysqlconn
+
+import (
+	"crypto/sha1"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/youtube/vitess/go/sqldb"
+)
+
+// Connect creates a connection to a server.
+// It then handles the initial handshake.
+//
+// If context is canceled before the end of the process, this function
+// will return nil, ctx.Err().
+//
+// FIXME(alainjobart) once we have more of a server side, add test cases
+// to cover all failure scenarios.
+func Connect(ctx context.Context, params *sqldb.ConnParams) (*Conn, error) {
+	netProto := "tcp"
+	addr := ""
+	if params.UnixSocket != "" {
+		netProto = "unix"
+		addr = params.UnixSocket
+	} else {
+		addr = net.JoinHostPort(params.Host, fmt.Sprintf("%v", params.Port))
+	}
+
+	// Figure out the character set we want.
+	characterSet, err := parseCharacterSet(params.Charset)
+	if err != nil {
+		return nil, err
+	}
+
+	// Start a background connection routine.  It first
+	// establishes a network connection, saves it into 'c', then
+	// works on the handshake. 'mu' protects 'c', so if the
+	// context is canceled we can safely read 'c'.
+	done := make(chan error, 1)
+	mu := sync.Mutex{}
+	var c *Conn
+	go func() {
+		var err error
+		var conn net.Conn
+
+		// Cap the Dial time with the context deadline, plus a
+		// few seconds. We want to reclaim resources quickly
+		// and not let this go routine stuck in Dial forever.
+		//
+		// We add a few seconds so we detect the context is
+		// Done() before timing out the Dial. That way we'll
+		// return the right error to the client (ctx.Err(), vs
+		// DialTimeout() error).
+		if deadline, ok := ctx.Deadline(); ok {
+			timeout := deadline.Sub(time.Now()) + 5*time.Second
+			conn, err = net.DialTimeout(netProto, addr, timeout)
+		} else {
+			conn, err = net.Dial(netProto, addr)
+		}
+		if err != nil {
+			done <- sqldb.NewSQLError(CRConnectionError, "", "net.Dial(%v,%v) failed: %v", netProto, addr, err)
+			return
+		}
+
+		// If we were canceled while connecting, stop right here.
+		// No need to write an error to done, as the other code path
+		// will also notice context cancelation.
+		select {
+		case <-ctx.Done():
+			conn.Close()
+			return
+		default:
+		}
+
+		mu.Lock()
```

Comment:
There's still a race here. If the context closes at this point, it can trigger the outer goroutine which can return without closing `c`, which would get eventually assigned here and the connection will remain open forever.